File: capy/src/e2e-cfun/js/dev/capylang/test/ReflectionFeature.test.js
const test = require('node:test');
const { assert, generatedModule } = require('./_support');

test('ReflectionFeature', () => {
    const reflection = generatedModule('dev/capylang/test/ReflectionFeature.js');
    const reflectedA = reflection.reflectionDataA(new reflection.A({ name: 'letter-a', a: 1 }));
    assert.equal(reflectedA.name, 'A');
    assert.deepEqual(reflectedA.fields.map(field => field.name), ['name', 'a']);
    assert.deepEqual(reflectedA.fields.map(field => field.value), ['letter-a', 1]);
    const reflectedB = reflection.reflectionDataB(new reflection.B({ name: 'letter-b', b: ['red', 'blue'] }));
    assert.deepEqual(reflectedB.fields.map(field => field.value), ['letter-b', ['red', 'blue']]);
    const reflectedEmpty = reflection.reflectionDataEmpty();
    assert.equal(reflectedEmpty.name, 'ReflectedEmpty');
    assert.equal(reflectedEmpty.packagePath, 'dev/capylang/test/ReflectionFeature');
    assert.deepEqual(reflectedEmpty.fields, []);
    const employee = reflection.reflectionDataEmployee(new reflection.ReflectedEmployee({ id: 'E-1', department: 'R&D', active: true }));
    assert.deepEqual(employee.fields.map(field => field.name), ['id', 'department', 'active']);
    assert.deepEqual(employee.fields.map(field => field.value), ['E-1', 'R&D', true]);
    const singleton = reflection.reflectionData(reflection.ReflectedSingleton);
    assert.equal(singleton.name, 'ReflectedSingleton');
    assert.deepEqual(singleton.annotations.map(annotation => annotation.name), ['ReflectionDataMarker']);
    assert.equal(annotationArgument(singleton.annotations[0], 'label').value.value, 'singleton');
    const json = reflection.reflectionJsonShape(new reflection.ReflectedEnvelope({
        label: 'outer',
        payload: new reflection.A({ name: 'inner', a: 7 }),
    }));
    assert.equal(json.value.get('label').value, 'outer');
    assert.equal(json.value.get('payload').value.get('name').value, 'inner');
    assert.equal(json.value.get('payload').value.get('a').value, 7);
});

test('reflects functional data and field annotations', () => {
    const reflection = generatedModule('dev/capylang/test/ReflectionFeature.js');
    const reflected = reflection.reflectionDataAnnotatedUser(new reflection.ReflectedAnnotatedUser({
        id: 'U-1',
        display: 'Ada',
    }));

    assert.deepEqual(reflected.annotations.map(annotation => annotation.name), ['ReflectionDataMarker']);
    assert.equal(annotationArgument(reflected.annotations[0], 'label').value.value, 'user');
    assert.equal(annotationArgument(reflected.annotations[0], 'label').value.kind, 'string');
    assert.equal(annotationArgument(reflected.annotations[0], 'order').value.value, 7);
    assert.equal(annotationArgument(reflected.annotations[0], 'active').value.value, true);
    assert.equal(reflected.pkg.path, 'dev/capylang/test/ReflectionFeature');

    assert.deepEqual(reflected.fields.map(field => field.annotations[0].name), ['ReflectionFieldMarker', 'ReflectionFieldMarker']);
    assert.equal(annotationArgument(reflected.fields[0].annotations[0], 'value').value.value, 'identifier');
    assert.equal(annotationArgument(reflected.fields[1].annotations[0], 'value').value.value, 'display');
    assert.deepEqual(reflected.fields.map(field => field.type.name), ['String', 'String']);
});

function annotationArgument(annotation, name) {
    return annotation.arguments.find(argument => argument.name === name);
}

File: capy/src/e2e-cfun/py/dev/capylang/test/reflection_feature_test.py
import os
import pathlib
import sys
import unittest


generated_dir = pathlib.Path(os.environ["CAPY_E2E_CFUN_PY_GENERATED_DIR"])
sys.path.insert(0, str(generated_dir))

import dev.capylang.test.ReflectionFeature as reflection


def annotation_argument(annotation, name):
    return next(argument for argument in annotation.arguments if argument.name == name)


class ReflectionFeaturePythonE2ETest(unittest.TestCase):
    def test_reflection_returns_data_value_fields(self):
        reflected = reflection.reflectionDataA(reflection.A({"name": "letter-a", "a": 1}))

        self.assertEqual(reflected.name, "A")
        self.assertEqual([field.name for field in reflected.fields], ["name", "a"])
        self.assertEqual([field.value for field in reflected.fields], ["letter-a", 1])

    def test_reflection_returns_functional_data_and_field_annotations(self):
        reflected = reflection.reflectionDataAnnotatedUser(reflection.ReflectedAnnotatedUser({
            "id": "U-1",
            "display": "Ada",
        }))

        self.assertEqual([annotation.name for annotation in reflected.annotations], ["ReflectionDataMarker"])
        self.assertEqual(annotation_argument(reflected.annotations[0], "label").value.kind, "string")
        self.assertEqual(annotation_argument(reflected.annotations[0], "label").value.value, "user")
        self.assertEqual(annotation_argument(reflected.annotations[0], "order").value.value, 7)
        self.assertEqual(annotation_argument(reflected.annotations[0], "active").value.value, True)
        self.assertEqual(reflected.pkg.path, "dev/capylang/test/ReflectionFeature")

        self.assertEqual(
            [field.annotations[0].name for field in reflected.fields],
            ["ReflectionFieldMarker", "ReflectionFieldMarker"],
        )
        self.assertEqual(annotation_argument(reflected.fields[0].annotations[0], "value").value.value, "identifier")
        self.assertEqual(annotation_argument(reflected.fields[1].annotations[0], "value").value.value, "display")
        self.assertEqual([field.type.name for field in reflected.fields], ["String", "String"])
